### PR TITLE
Use Result for directions response

### PR DESCRIPTION
**Pull request by JThramer**
Using `Foundation.Result` as the base type returned in `Directions.calculate`. Implements #414. Thanks @SebastianOsinski!

To-Do: 
- [x] Finish fixing tests
- [x] Update docs

/cc @mapbox/navigation-ios

### CODE REVIEW

File: Sources/MapboxDirections/Directions.swift [hunk 1]
```diff
@@ -57,14 +57,22 @@ let userAgent: String = {
  Each result produced by the directions object is stored in a `Route` object. Depending on the `RouteOptions` object you provide, each route may include detailed information suitable for turn-by-turn directions, or it may include only high-level information such as the distance, estimated travel time, and name of each leg of the trip. The waypoints that form the request may be conflated with nearby locations, as appropriate; the resulting waypoints are provided to the closure.
  */
 open class Directions: NSObject {
+    
+    public typealias MapMatchingError = (response: MapMatchingResponse, error: DirectionsError)
+    
+//    public typealias RouteError: Error = (result: RouteResponse, error: DirectionsError)
+//    public typealias MapMatchingError: Error = (result: MapMatchingResponse, error: DirectionsError)
```
Comment (JThramer) [inline Sources/MapboxDirections/Directions.swift]: remove comment

File: Sources/MapboxDirections/Directions.swift [hunk 2]
```diff
@@ -57,14 +57,19 @@ let userAgent: String = {
...
+    public typealias MapMatchingError = (response: MapMatchingResponse, error: DirectionsError)
+    
+    public typealias DirectionsSession = (options: DirectionsOptions, credentials: DirectionsCredentials)
+    
     /**
      A closure (block) to be called when a directions request is complete.
      
      - parameter response: A `RouteResponse` object that contains the entire payload of the Directions API solution. See `RouteResponse.swift` for more information.
      
      - parameter error: The error that occurred, or `nil` if the solution was obtained successfully.
```
Comment (Udumft) [inline Sources/MapboxDirections/Directions.swift]: Please do not forget to update docs in result

File: Sources/MapboxDirections/Directions.swift [hunk 3]
```diff
@@ -152,40 +155,36 @@ open class Directions: NSObject {
...
+                            completionHandler(session, .failure(apiError))
                         }
                         return
                     }
                     
                     let result = try decoder.decode(RouteResponse.self, from: data)
                     guard result.routes != nil else {
                         DispatchQueue.main.async {
-                            completionHandler(result, .unableToRoute)
+                            completionHandler(session, .failure(.unableToRoute))
```
Comment (Udumft) [inline Sources/MapboxDirections/Directions.swift]: I am not sure about original reasons, but as @1ec5 mentioned in [original proposal](https://github.com/mapbox/mapbox-directions-swift/issues/414#issuecomment-586404352),  this was the case which returned an error and a waypoints at the same time. I don't know if it was done for a good reason or not, but apparently such behavior will be lost.

File: Sources/MapboxDirections/Directions.swift [hunk 4]
```diff
@@ -57,14 +57,19 @@ let userAgent: String = {
  Each result produced by the directions object is stored in a `Route` object. Depending on the `RouteOptions` object you provide, each route may include detailed information suitable for turn-by-turn directions, or it may include only high-level information such as the distance, estimated travel time, and name of each leg of the trip. The waypoints that form the request may be conflated with nearby locations, as appropriate; the resulting waypoints are provided to the closure.
  */
 open class Directions: NSObject {
+    
+    public typealias MapMatchingError = (response: MapMatchingResponse, error: DirectionsError)
```
Comment (Udumft) [inline Sources/MapboxDirections/Directions.swift]: Unused synonym. Did you intend to use it as a result type at some point?

**Comment by JThramer** [comment]
Blocked from merge by https://github.com/mapbox/mapbox-directions-swift/pull/418, need to fix cartfile

File: Sources/MapboxDirections/Directions.swift (hunk 3 above)
Comment (JThramer) [reply Sources/MapboxDirections/Directions.swift:174]: as it turns out, `session` contains an options object with the waypoints.

File: Tests/MapboxDirectionsTests/DirectionsTests.swift [hunk 5]
```diff
@@ -113,16 +117,19 @@ class DirectionsTests: XCTestCase {
         
         let directions = Directions(credentials: BogusCredentials)
         let opts = RouteOptions(locations: [one, two])
-        directions.calculate(opts, completionHandler: { (response, error) in
+        directions.calculate(opts, completionHandler: { (session, result) in
             expectation.fulfill()
```
Comment (Udumft) [inline Tests/MapboxDirectionsTests/DirectionsTests.swift:121]: fulfilling expectation before other checks seems incorrect...

File: Tests/MapboxDirectionsTests/DirectionsTests.swift [hunk 6]
```diff
@@ -157,20 +164,20 @@ class DirectionsTests: XCTestCase {
...
```
Comment (Udumft) [inline Tests/MapboxDirectionsTests/DirectionsTests.swift:168]: Same fullfilment not as above

File: Tests/MapboxDirectionsTests/DirectionsTests.swift (hunk 5 above)
Comment (JThramer) [reply Tests/MapboxDirectionsTests/DirectionsTests.swift:121]: Why do you think that? Keep in mind that the fulfillment is not there to say "okay test is over", It's merely there to ensure that the (faked) directions request takes no longer than two seconds to call it's `completionHandler.`

Comment (JThramer) [reply Tests/MapboxDirectionsTests/DirectionsTests.swift:121]: I was all ready to explain why you're wrong, but as it turns out, you're not. 

Very good catch. 🙇‍♂️

**Pull request merged**